clang/labs/lab1/kashirin_alexander/Rename_Plugin.cpp

```diff
@@ -0,0 +1,211 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+#include "clang/Lex/Lexer.h"
+#include "clang/Rewrite/Core/Rewriter.h"
+
+enum class IdType { Var, Func, Class };
+
+class RenameVisitor : public clang::RecursiveASTVisitor<RenameVisitor> {
+public:
+  bool isValidIdentifier(const std::string &name) const {
+    clang::Lexer lexer(clang::SourceLocation(), clang::LangOptions(),
+                       name.c_str(), name.c_str(),
+                       name.c_str() + name.length());
+    for (clang::Token tok; !lexer.LexFromRawLexer(tok);)
+      if (tok.isNot(clang::tok::identifier))
+        return false;
+    return true;
+  }
+
+  explicit RenameVisitor(clang::Rewriter rewriter, IdType type,
+                         clang::StringRef cur_name, clang::StringRef new_name)
+      : rewriter(rewriter), type(type), cur_name(cur_name), new_name(new_name) {
+  }
+
+  bool VisitFunctionDecl(clang::FunctionDecl *func) {
+    if (type == IdType::Func && func->getName() == cur_name &&
+        isValidIdentifier(new_name)) {
+      rewriter.ReplaceText(func->getNameInfo().getSourceRange(), new_name);
+    }
+    return true;
+  }
+
+  bool VisitCallExpr(clang::CallExpr *call) {
+    if (type == IdType::Func) {
+      clang::FunctionDecl *callee = call->getDirectCallee();
+      if (callee && callee->getName() == cur_name) {
+        rewriter.ReplaceText(call->getCallee()->getSourceRange(), new_name);
```

Comment:
Missed validity check for "new_name". Here and in a couple of other places. It makes sense to move it to the place where you parse cmd line args got from user.